[expo] proxy NativeModules to add better error messaging in dev clients

esamelson opened the pull request:
# Why

ENG-5782

This PR adds a JS `Proxy` around react-native's `NativeModules` export, which allows us to log a descriptive and helpful error message when developers try to access unsupported native modules in a development client, such as Expo Go.

#18064 was a start in this direction, but was pretty limited in the cases it could catch. This PR lets us catch virtually any case where someone tries to access an unsupported module.

This is still a work in progress -- in particular, this PR doesn't include a way to disable this error message, either globally or per-module -- but I'm hoping to get some eyes on this approach before going much further.

# How

- Use metro's `resolveRequest` to intercept the `require('./Libraries/BatchedBridge/NativeModules')` statement in `react-native/index.js`, which allows us to replace the `NativeModules` object right before it's exported by the library.
- Wrap it in a Proxy with a `get` handler; this allows us to intercept all attempts to access properties on the object.
- If the property doesn't exist and we're in the correct environment (Expo Go or a development build) log a descriptive error with the module name and steer people to development builds.

# Test Plan

I've been testing this locally with `react-native-ble-manager` and Expo Go. This is what shows up in my logs when I import the library:

<img width="1132" alt="Screen Shot 2022-08-02 at 2 54 22 PM" src="https://user-images.githubusercontent.com/19958240/182481394-eaff2785-6a10-49be-b534-cb10f8f3c335.png">

# Risks / drawbacks

- There is no way to differentiate between someone checking for the existence of a native module vs. actually trying to use it, which means there will definitely be some false positives, especially in modules with resilient JS code. I don't have a sense of how widespread these false positives will be. My sense is that some are ok as long as we provide a way to suppress these errors on a per-module basis (not yet implemented). Curious what others think.
- When this is enabled, the `NativeModules` export from react-native becomes a `Proxy` rather than a regular object. So far I haven't found any ways in which its behavior differs in a consequential and unintended way, but it's possible there are some edge cases. (Things like `console.log(NativeModules)` fail in slightly different ways than before.)
- This will not catch any attempts to access nested native modules, such as those on e-m-c's `NativeModulesProxy` (but those are already handled by #18064).
- Some additional maintenance cost (e.g. will break if RN changes the way it imports NativeModules internally).

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

esamelson commented on packages/expo/src/proxies/NativeModules.ts:
```diff
@@ -0,0 +1,52 @@
+    get(target, prop) {
+      const value = target[prop];
+      if (
+        (value === null || value === undefined) &&
```
TIL! Though I actually think the first way makes the intent of the code clearer. If I were to come across the second one I might just assume the `==` was a typo.

esamelson commented on packages/expo/src/proxies/NativeModules.ts:
```diff
@@ -0,0 +1,52 @@
+      ) {
+        hasWarnedForModule.push(prop.toString());
+
+        const isRunningInStoreClient =
+          global.ExpoModules?.NativeModulesProxy?.modulesConstants?.ExponentConstants
+            ?.executionEnvironment === 'storeClient' ||
+          target.NativeUnimoduleProxy?.modulesConstants?.ExponentConstants?.executionEnvironment ===
+            'storeClient';
```
Hmm... this works, but I do get a (warranted) require cycle warning. Not sure if it's worth trying to suppress that 😬


<img width="384" alt="Screen Shot 2022-08-04 at 5 09 26 PM" src="https://user-images.githubusercontent.com/19958240/182975551-01f921e0-f180-4212-9296-f928b7b561f8.png">

brentvatne commented:
this is an intriguing idea!

- it won't cover turbomodules in bridgeless mode (https://github.com/facebook/react-native/blob/main/Libraries/TurboModule/TurboModuleRegistry.js#L15) - this is fine for now
- our metro config currently isn't strictly required unless you're using expo-updates (to inject asset hash) but in practice most apps should be using it, so this may be fine
- is there any way we could make this cleaner by upstreaming some changes? for example, i think depending on overriding the native module at the resolver level is potentially a bit fragile. possibly react-native would allow us to add an api to inject a wrapper. i'm not opposed to shipping works-for-today solution and trying to upstream something for an upcoming release. now that we track react-native more closely and releases to react-native are more frequent, this isn't such a stretch.
- react-native already uses a native module proxy when using jsi: https://github.com/facebook/react-native/blob/3f8071d638d7c9cd25c0eb7f6c2169c776e271ea/Libraries/BatchedBridge/NativeModules.js#L176-L177 / https://github.com/facebook/react-native/blob/521011d4cc713dfce97dc8872fd0f5171e587b5d/ReactCommon/jsiexecutor/jsireact/JSIExecutor.cpp#L83-L89. an upside of this is that i don't think people depend on anything in the public api for NativeModules beyond being able to access properties on the object

Kudo commented:
i don't really like to replace the NativeModules export and try to seek for other possibilities. these's one maybe worth to take a look.
#18064's approach makes sense to me. is the problem that we may not catch all cases for the matched strings like `Native module cannot be null`? if so, react-native provides a [NativeModulePerfLogger when failed to require a module](https://github.com/facebook/react-native/blob/3f8071d638d7c9cd25c0eb7f6c2169c776e271ea/ReactCommon/reactperflogger/reactperflogger/NativeModulePerfLogger.h#L80). it is sent from [`global.nativeModuleProxy["something"]` is null](https://github.com/facebook/react-native/blob/3f8071d638d7c9cd25c0eb7f6c2169c776e271ea/ReactCommon/jsiexecutor/jsireact/JSINativeModules.cpp#L49)
we could try to enable the logger and catch the `moduleJSRequireEndingFail` error. one down side is that the approach doesn't support remote debugging mode (proxy bridge mode).

EvanBacon commented on packages/@expo/metro-config/proxies/NativeModules.js:
```diff
@@ -0,0 +1,12 @@
+  const { createProxyForNativeModules } = require('expo/build/proxies/NativeModules');
+  proxy = createProxyForNativeModules(NativeModules);
+} catch {
+  // do nothing, this is expected if no supported version of the expo package is installed
```
This seems like not a great idea, we probably want to make use of all error messages and not introduce random error hiding.

EvanBacon commented:
1. We use `resolveRequest` for [web support here](https://github.com/expo/expo/blob/46dfeb9dfbeb8bb5de72733baf4268605935b03c/packages/%40expo/cli/src/start/server/metro/withMetroMultiPlatform.ts#L102).
2. The `resolveRequest` interface is different (no fourth property) in later versions of Metro (possibly the version we're currently on). https://facebook.github.io/metro/docs/configuration/#resolverequest
3. I've played with this same concept for native import support, but it happens during the transformation: 
  - Metro works like: resolve (yours is here) -> transform (mine was here) -> serialize. [Reference in case this helps inspire](https://github.com/EvanBacon/expo-quick-actions/blob/6e152d54da74799a19487a3648bf3ee9d5c28e22/example/transformer/templates.ts#L15).
4. I'd recommend testing this well, both in the config layer and the application code layer.
5. Don't worry too much about the treating out metro config as optional, I can't think of a JS framework that has an optional bundler configuration, it's too limiting. We require the use of `expo/metro-config` for automatically using babel-preset-expo, setting up web support, symbolication that accounts for the `expo` logger, exotic mode, and once [`require.context`](https://github.com/facebook/metro/pull/822/files) lands we will be enabling it by default for [routing](https://linear.app/expo/project/auto-navigation-50ddb36bec78). That said, we should always opt to ensure our bundler config can be extended so as not to limit what the user can do in the bundler space.

esamelson committed:
working new implementation based on Kudo's idea

esamelson committed:
add a way to disable globally or for certain modules

esamelson commented:
Thanks for the comments and ideas, everyone. I've updated this to a simpler implementation suggested by @Kudo , which uses `Object.defineProperty` and doesn't need to rely on metro at all 🎉 It also allows us to export a simple function, `disableMissingNativeModuleErrors` that can be called at the top of App.js to disable these errors globally or on a per-module basis.

I also switched from console.error to throwing, which will allow things like
```
try {
   BleManager = require('react-native-ble-manager');
} catch (e) { ... }
```
for anyone who would rather catch the errors than disable them.

I've tested this in debug & release Expo Go, and in a dev client, with the following modules: `react-native-ble-manager`, `react-native-vision-camera`, `@react-native-firebase/database`, `react-native-plaid-link-sdk` and it works as expected with all of them. I've also tried things like `console.log(NativeModules)`, `JSON.stringify(NativeModules)`, `NativeModules.toString()` and ensured that the behavior doesn't differ at all when the proxy is applied. Reloading / fast refresh also seems to work as expected. Happy to test any other scenarios anyone suggests, as well.

Kudo commented on packages/expo/src/proxies/NativeModules.ts:
```diff
@@ -0,0 +1,80 @@
+        // only want to throw once per module
+        !hasWarnedForModule.includes(prop.toString())
+      ) {
+        hasWarnedForModule.push(prop.toString());
```
nit: maybe we could store `const propName = prop.toString();` and reuse it. there're are six `prop.toString()` here.

Kudo reviewed (approved):
just leaving a nit comment. the changes mostly look good to me.

esamelson committed:
feedback: store propName

brentvatne commented:
this looks great! the only thing missing in this pr is we should ensure this is covered in our docs as well. we can say it's only available in sdk 47+ (even though it's not yet released, this is what we have been doing recently and i find it works fine!)

brentvatne commented:
oh one last thing - let's test this in hermes as well, to ensure that there is nothing different about the proxy implementation there that will cause issues

ide commented on packages/expo/src/Expo.fx.tsx:
```diff
@@ -82,4 +83,11 @@ if (__DEV__) {
+
+  const proxiedNativeModules = createProxyForNativeModules(ReactNative.NativeModules);
+  Object.defineProperty(ReactNative, 'NativeModules', {
+    get() {
+      return proxiedNativeModules;
+    },
```
Preserve extensibility:
```suggestion
    get() {
      return proxiedNativeModules;
    },
    configurable: true,
    enumerable: true,
```

ide commented on packages/expo/src/proxies/NativeModules.ts:
```diff
@@ -0,0 +1,81 @@
+import { ExecutionEnvironment } from 'expo-constants';
+
+const PROPS_TO_IGNORE: string[] = [
+  /**
+   * We don't want to throw when the expo or expo-modules-core packages try to access any of these
+   * modules, since they have built-in fallbacks.
+   */
+  'DevLoadingView',
+  'EXDevLauncher',
+  'EXReactNativeEventEmitter',
+  'NativeUnimoduleProxy',
+  /**
+   * Other methods that can be called on the NativeModules object that we should ignore. The
+   * underlying NativeModules object is sometimes a proxy itself so may not have these methods
+   * defined.
+   */
+  'hasOwnProperty',
+  'isPrototypeOf',
+  'toJSON',
+  'toLocaleString',
+  'toString',
+  'valueOf',
+  '$$typeof',
+];
+const hasWarnedForModule: string[] = [];
+
+let additionalModulesToIgnore: string[] = [];
```
I believe all supported JS interpreters now implement Set

ide commented on packages/expo/src/proxies/NativeModules.ts:
```diff
@@ -0,0 +1,81 @@
+  });
+}
+
+export function disableMissingNativeModuleErrors(moduleNames?: string[] | string) {
```
Document what this does, including the behavior that it sets all of the modules to disable warnings for, rather than adding more to the list.

ide commented on packages/expo/src/proxies/NativeModules.ts:
```diff
@@ -0,0 +1,81 @@
+  return new Proxy(NativeModules, {
+    get(target, prop) {
+      const value = target[prop];
+      const propName = prop.toString();
```
I don't think it is safe to stringify the property name since it is legitimate to index into an object using a Symbol. The rough logic should be: if the property name is a symbol or the checks are disabled, return `Reflect.get(target, prop, receiver)`. Reflect is designed for forwarding proxy methods to the underlying object.

Once Symbols are taken care of, it's guaranteed the property name is a string.

ide commented on packages/expo/src/proxies/NativeModules.ts:
```diff
@@ -0,0 +1,81 @@
+        hasWarnedForModule.push(propName);
+
+        const isRunningInStoreClient =
+          global.ExpoModules?.NativeModulesProxy?.modulesConstants?.ExponentConstants
+            ?.executionEnvironment === ExecutionEnvironment.StoreClient ||
+          target.NativeUnimoduleProxy?.modulesConstants?.ExponentConstants?.executionEnvironment ===
+            ExecutionEnvironment.StoreClient;
```
Are all these question marks necessary?

esamelson committed:
feedback

esamelson committed:
changelog

esamelson commented:
> oh one last thing - let's test this in hermes as well, to ensure that there is nothing different about the proxy implementation there that will cause issues

thanks for the suggestion @brentvatne -- ran the same tests in hermes (dev client on iOS, Expo Go on Android) and everything still behaved as expected.

> the only thing missing in this pr is we should ensure this is covered in our docs as well.

I'll do this in a separate followup PR after this one lands, if that sounds ok. Have one other cleanup task to do as well, made linear tasks for both (ENG-6097 and ENG-6098)

brentvatne commented:
sounds good @esamelson!

ide commented on packages/expo/src/proxies/NativeModules.ts:
```diff
@@ -0,0 +1,90 @@
+   * defined.
+   */
+  'hasOwnProperty',
+  'isPrototypeOf',
```
Another method to add:
```suggestion
  'isPrototypeOf',
  'propertyIsEnumerable',
```

esamelson replied:
Good catch, thanks

esamelson committed:
add propertyIsEnumerable to list of exceptions

esamelson committed:
change method for disabling errors

esamelson committed:
update message to point to expo.fyi page
